File: subarray_sum_equals_k/Subarray_Sum_Equals_K.cpp
class Subarray_Sum_Equals_K {
public:
    int subarraySum(vector<int>& nums, int k) {
        if (nums.empty()) {
            return 0;
        }
        
        unordered_map<int, int> table;
        table[0] = 1;
        int prefix_sum = 0;
        int num_subarray_sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            prefix_sum += nums[i];
            if (table.find(prefix_sum - k) != table.end()) {
                num_subarray_sum += table[prefix_sum - k];
            }
            table[prefix_sum]++;
        }
        
        return num_subarray_sum;
    }
};

File: subarray_sum_equals_k/SubarraySumEqualsK.java
class SubarraySumEqualsK {
    public int subarraySum(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        
        int prefixSum = 0;
        int numSubarray = 0;
        Map<Integer, int[]> table = new HashMap<>();
        table.put(prefixSum, new int[]{1});
        for (int i = 0; i < nums.length; i++) {
            prefixSum += nums[i];
            if (table.containsKey(prefixSum - k)) {
                numSubarray += table.get(prefixSum - k)[0];
            }
            if (!table.containsKey(prefixSum)) {
                table.put(prefixSum, new int[]{0});
            }
            table.get(prefixSum)[0]++;
        }
        
        return numSubarray;
    }
}